### PR TITLE
Add generic SASL bind support using rsasl

### PR DESCRIPTION
**Pull request by dequbed**
Tag, you're it!

Anyway, I kinda sorta need the ability to do SASL bind for unrelated related reasons. So, based on the famous (and always wrongly attributed to Mahatma Gandhi) quote "be the change you wish to see — Arleen Lorrance", here you go! :P

The main advantage of using rsasl here is that we get separation of concern; the ldap side of things does not have to change in any way to allow for any and all SASL mechanisms to be used. In fact it won't even have to change to support mechanisms that haven't even been invented yet, as rsasl makes those entirely abstracted modular plugins that neither of us needs to care about. ¯\\\_(ツ)\_/¯

This separation also extends to selection of mechanisms, getting required information from the user for them, and all of that jazz, the `Arc<SASLConfig>` (which is wrapped in an `Arc` because it's just a fancy trait object, in other words unsized) is a freshly-fashioned configuration object provided by the downstream user that lets said user configure all the mechanism selection, preferences and all the authentication data required for the authentication.

I mean, I did write rsasl so we don't get to use other people's work here, but still, separation of concerns is good, and modular software doubly so.

Anyway, rsasl makes *heavy* use of feature unification to remove as much code as possible when it's not used, so in this configuration with only "provider" selected rsasl is *tiny*. This is useful as especially with channel binding and security layers rsasl will need to be pretty deeply integrated into the remaining code, making conditional compilation really painful really fast. That being said, if you prefer it as an optional dependency we can do that too.

Lastly:

Compared to the existing GSS-API bind this PR doesn't implement security layers because … well mostly because I haven't decided how I want to handle security layers on the rsasl side yet. They do kinda stink and I'd love to just … *not*, but I fear that won't be a good option either.

Also this MR does not yet do channel binding. I've been meaning to write some utilities to extract the required channel binding data from rustls and native-tls but I've been dreading it and haven't yet gotten around to it. (I really should though <.<)

### CODE REVIEW

**Comment by inejge** [comment]
Thanks for the PR and sorry about the delay. Briefly, I don't think I'd merge this into the main branch, but it can be kept and maintained in a parallel branch until it's a bit more fleshed out. Expect a more detailed review shortly.

**Comment by inejge** [comment]
OK, I _finally_ got some time to write a more detailed response. As I wrote in the first reply, I wouldn't merge the PR into the master branch. The principal reason is that my general impression of rsasl is that it's still rather incomplete/experimental, and I don't want to mix experimental features with production code, or give an impression that they're more usable than they actually are -- and from long experience I know that documenting something as incomplete has never stopped people from using it anyway and then complaining when it doesn't work as expected.

For incorporating any code in the library, I want it to be a) as complete as possible, and b) portable to the big three non-mobile operating systems, with minimal loss of functionality. That's why I never merged SPNEGO/NTLM auth; at the time, it couldn't use Windows session credentials, nor could it deal with security layers or channel bindings. I think it still doesn't handle the former, but I'm following it with interest, since it promises Rust-native Kerberos handling among other things. To circle back to rsasl, it can't deal with GSSAPI as it is; you've also mentioned lack of security layers/channel bindings yourself, so you're perfectly aware of that deficiency. For the main crate, I must have all of it.

Another thing that is tolerable in experimental code, and which I wouldn't incorporate in the main branch, is no-questions-asked querying of Root DSE in the SASL binding code: for production use, it would have to be cached somehow, and I'm not sure that such queries belong to the library itself; a level above, perhaps. (One of my design principles for the library is to provide mechanisms and minimize policy decisions.)

The example in the PR uses a fictitious connection URL. Again, that's OK for the separate branch, but master examples have to work against OpenLDAP whose configuration is in the `data` subdirectory.

To summarize, this means that I'll merge the PR into the `rsasl` branch for now. I don't believe that SASL has great future in LDAP clients (it will stay confined to EXTERNAL and GSSAPI/SPNEGO; I'm very curious about your need for SCRAM-SHA-256), but I think that, eventually, rsasl could replace my ad-hoc hacks, just like `sspi` is a potential replacement for C Kerberos libraries. So I'm open to keeping it in-tree to follow the development. Periodic rebases will probably be needed to keep the code up with master, but I don't expect any drastic changes to the main codebase.
